drivers/hmis/spec/models/hmis/form/definition_validation_spec.rb

```diff
@@ -0,0 +1,192 @@
+  let(:default_args) { { entry_date: entry_date, exit_date: exit_date } }
+
+  describe 'find and validate assessment' do
+    it 'should error if assessment date value is missing from hud_values' do
+      HUD_ASSESSMENT_ROLES.each do |role|
+        definition = Hmis::Form::Definition.find_by(role: role)
+        date, errors = definition.find_and_validate_assessment_date(hud_values: { 'foo': 'bar' }, **default_args)
+        expect(date).to be_nil
+        expect(errors.length).to eq(1)
+        expect(errors[0].type).to eq(:required)
+        expect(errors[0].attribute).to eq(definition.assessment_date_item.field_name)
+      end
+    end
+
+    it 'should error if assessment date value is null in hud_values' do
+      HUD_ASSESSMENT_ROLES.each do |role|
+        definition = Hmis::Form::Definition.find_by(role: role)
+        link_id = definition.assessment_date_item.link_id
+        date, errors = definition.find_and_validate_assessment_date(hud_values: { link_id => nil }, **default_args)
+        expect(date).to be_nil
+        expect(errors.length).to eq(1)
+        expect(errors[0].type).to eq(:required)
+        expect(errors[0].attribute).to eq(definition.assessment_date_item.field_name)
+      end
+    end
+
+    it 'should succeed if assessment date is the same as entry date for all roles' do
+      HUD_ASSESSMENT_ROLES.each do |role|
+        definition = Hmis::Form::Definition.find_by(role: role)
+        link_id = definition.assessment_date_item.link_id
+        assessment_date = entry_date.strftime('%Y-%m-%d')
+
+        date, errors = definition.find_and_validate_assessment_date(hud_values: { link_id => assessment_date }, **default_args)
+        expect(errors).to be_empty
+        expect(date).to eq(entry_date)
+      end
+    end
+
+    it 'should error if assessment date is before entry date (non-intake)' do
+      HUD_ASSESSMENT_ROLES.excluding(:INTAKE).each do |role|
+        definition = Hmis::Form::Definition.find_by(role: role)
+        link_id = definition.assessment_date_item.link_id
+        assessment_date = (entry_date - 1.day).strftime('%Y-%m-%d')
+
+        date, errors = definition.find_and_validate_assessment_date(hud_values: { link_id => assessment_date }, **default_args)
+        expect(errors.length).to eq(1)
+        expect(errors[0].type).to eq(:out_of_range)
+        expect(errors[0].attribute).to eq(definition.assessment_date_item.field_name)
+        expect(date).to be_nil
+      end
+    end
+
+    it 'should succeed if assessment date is before entry date (intake)' do
+      definition = Hmis::Form::Definition.find_by(role: 'INTAKE')
+      link_id = definition.assessment_date_item.link_id
+      assessment_date = (entry_date - 1.day).strftime('%Y-%m-%d')
+
+      date, errors = definition.find_and_validate_assessment_date(hud_values: { link_id => assessment_date }, **default_args)
+      expect(date).to eq(entry_date - 1.day)
+      expect(errors).to be_empty
```

Comment:
Should this also set the entry date to be the assessment_date?  If so, let's check that.

Reply:
This function `find_and_validate_assessment_date` does not actually set the date on any record. It just returns it.

I should add a test case for that to `submit_assessment_spec.rb` which currently only tests on Update assessments, not Intakes or Exits.